stko/optimizers/xtb.py

```diff
@@ -0,0 +1,509 @@
+                stdout=f,
+                stderr=sp.PIPE,
+                # Shell is required to run complex arguments.
+                shell=True
+            )
+
+    def _run_optimizations(self, mol):
+        """
+        Run loop of optimizations on `mol` using xTB.
+
+        Parameters
+        ----------
+        mol : :class:`.Molecule`
+            The molecule to be optimized.
+
+        Returns
+        -------
+        mol : :class:`.Molecule`
+            The optimized molecule.
+
+        opt_complete : :class:`bool`
+            Returns ``True`` if the calculation is complete and
+            ``False`` if the calculation is incomplete.
+
+        """
+
+        for run in range(self._max_runs):
+            xyz = f'input_structure_{run+1}.xyz'
+            out_file = f'optimization_{run+1}.output'
+            mol.write(xyz)
+            self._run_xtb(xyz=xyz, out_file=out_file)
+            # Check if the optimization is complete.
+            coord_file = 'xtbhess.coord'
+            coord_exists = os.path.exists(coord_file)
+            output_xyz = 'xtbopt.xyz'
+            opt_complete = self._is_complete(out_file)
+            if not opt_complete:
+                if coord_exists:
+                    # The calculation is incomplete.
+                    # Update mol from xtbhess.coord and continue.
+                    mol = mol.with_structure_from_file(coord_file)
+                else:
+                    # Update mol from xtbopt.xyz.
+                    mol = mol.with_structure_from_file(output_xyz)
+                    # If the negative frequencies are small, then GFN
+                    # may not produce the restart file. If that is the
+                    # case, exit optimization loop and warn.
+                    self.incomplete.add(mol)
+                    logging.warning(
+                        f'Small negative frequencies present in {mol}.'
+                    )
+                    return False
+            else:
+                # Optimization is complete.
+                # Update mol from xtbopt.xyz.
+                mol = mol.with_structure_from_file(output_xyz)
+                break
+
+        return mol, opt_complete
+
+    def optimize(self, mol):
+        """
+        Optimize `mol`.
+
+        Parameters
+        ----------
+        mol : :class:`.Molecule`
+            The molecule to be optimized.
+
+        Returns
+        -------
+        None : :class:`NoneType`
+
+        """
+
+        # Remove mol from self.incomplete if present.
+        if mol in self.incomplete:
+            self.incomplete.remove(mol)
+
+        if self._output_dir is None:
+            output_dir = str(uuid.uuid4().int)
+        else:
+            output_dir = self._output_dir
+        output_dir = os.path.abspath(output_dir)
+
+        if os.path.exists(output_dir):
+            shutil.rmtree(output_dir)
+
+        os.mkdir(output_dir)
+        init_dir = os.getcwd()
+        os.chdir(output_dir)
```

Comment:
Why do we change directory when optimising? Can we just write to a given path instead of changing directory? I feel less bugs are likely to happen this way.

Reply:
This is also probs more a long term change

Reply:
This will stay as it is for now.